[Site Isolation] getDisplayMedia does not work in a out of process iframes

bnham opened the pull request:
#### 1935a83b4fdaacb26d179f4189020bd39efcc04e
<pre>
[Site Isolation] getDisplayMedia does not work in a out of process iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=287299">https://bugs.webkit.org/show_bug.cgi?id=287299</a>
<a href="https://rdar.apple.com/129194055">rdar://129194055</a>

Reviewed by Charlie Wolfe and Eric Carlson.

Starting a display stream via getDisplayMedia does not work in an OOP iframe when site isolation is
enabled. This is because many of the related operations assumed that only the main frame&apos;s process
could start a capture.

Fix this by making various parts of setting up a display stream (such as `updateCaptureAccess`,
`UserMediaAccessWasGranted`, `UserMediaAccessWasDenied`, extending sandbox to allow capture, etc.)
to use the process associated with the frame in the UserMediaRequest, rather than unconditionally
using the main frame&apos;s process.

Since a WebProcessProxy associated with an OOP iframe can enter the process cache while the main
page is still has an active capture session, also make `WebProcessProxy::removeWebPage` check to see
if it&apos;s okay to remove capture sandbox extensions when pages are removed from the process.

* LayoutTests/http/tests/site-isolation/mediastream/getDisplayMedia-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getDisplayMedia-starts.html: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-audio-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-audio-starts.html: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-video-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-video-starts.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getDisplayMedia-starts-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getUserMedia-audio-starts-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getUserMedia-video-starts-frame.html: Added.
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::isPlayingMediaDidChange):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::revokeSandboxExtensionsIfNeededForPage):
(WebKit::UserMediaPermissionRequestManagerProxy::disconnectFromPage):
(WebKit::UserMediaPermissionRequestManagerProxy::captureDevicesChanged):
(WebKit::UserMediaPermissionRequestManagerProxy::denyRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::captureStateChanged):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/UserMediaProcessManager.cpp:
(WebKit::UserMediaProcessManager::willCreateMediaStream):
(WebKit::UserMediaProcessManager::revokeSandboxExtensionsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::userMediaPermissionRequestManagerIfExists):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::setOrientationForMediaCapture):
(WebKit::WebPageProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::removeWebPage):

Canonical link: <a href="https://commits.webkit.org/290715@main">https://commits.webkit.org/290715@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/933544585ded23823ebed4d8d7abc06b84206c4b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10368 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/45778 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95850 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/41624 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/10763 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/18685 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69845 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27379 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8189 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82341 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50185 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7963 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36732 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/40746 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37800 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/97824 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18026 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13221 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78867 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18285 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78065 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19292 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22529 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/21156 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/11334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18035 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23380 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/17774 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/19558 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15464,6 +15469,18 @@ Ref<WebNavigationState> WebPageProxy::protectedNavigationState()
+    if (auto* remotePage = m_browsingContextGroup->remotePageInProcess(*this, frame->process()))
+        return &remotePage->siteIsolatedProcess();
+    if (&frame->process() == m_legacyMainFrameProcess.ptr())
+        return m_legacyMainFrameProcess.ptr();
```
Why do you need these last two ifs? Can't you just return `frame->process()`?

bnham replied:
I was originally doing that (and just returning a `Ref<WebProcessProxy>` or `WebProcessProxy&`, but in case you pass in a frame ID that's not associated with the process, I wanted to return nullptr. Not sure if that makes sense or not.

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14747,7 +14750,9 @@ void WebPageProxy::gpuProcessExited(ProcessTerminationReason)
         auto& gpuProcess = configuration().processPool().ensureGPUProcess();
-        gpuProcess.updateCaptureAccess(activeAudioCapture, activeVideoCapture, activeDisplayCapture, m_legacyMainFrameProcess->coreProcessIdentifier(), identifier(), [] { });
+        forEachWebContentProcess([&](auto& webProcess, auto pageID) {
+            gpuProcess.updateCaptureAccess(activeAudioCapture, activeVideoCapture, activeDisplayCapture, m_legacyMainFrameProcess->coreProcessIdentifier(), identifier(), [] { });
```
Did you mean to replace `m_legacyMainFrameProcess`? Otherwise the loop isn't needed.

bnham replied:
Yeah this is a mistake, I meant to use `webProcess->coreProcessIdentifier()`

charliewolfe commented on Source/WebKit/UIProcess/UserMediaProcessManager.cpp:
```diff
@@ -188,20 +192,27 @@ bool UserMediaProcessManager::willCreateMediaStream(UserMediaPermissionRequestMa
+            if (!pages.contains(Ref { *proxyPage }))
+                return;
+            hasAudioCapture |= proxyPage->isCapturingAudio();
+            hasVideoCapture |= proxyPage->isCapturingVideo();
```
I think this will still be broken because, right now, any process can overwrite the media state set by another process.

In the future, we’ll likely need to move these functions out of `WebPageProxy` so we can track the media state set by each process on the page separately. Then we could also revoke sandbox extensions when an individual process no longer needs them, rather than waiting until all processes associated with a page no longer need them.

bnham replied:
Hm yeah, I didn't know how `Page::updateIsPlayingMedia` eventually changes state in the media capture flags in `WebPageProxy`, which seems like a pretty big problem.

bnham replied:
I suspect the minimal change would be to have each `RemotePageProxy` also track its own `mediaState`.

bnham replied:
But then there's also a bunch of related flags that need fixing, like `activityState` etc.

charliewolfe replied:
Why does `activityState` need fixing? From when I briefly looked it didn't seem like it was being updated by the WP.

charliewolfe commented on LayoutTests/http/tests/site-isolation/resources/getDisplayMedia-starts-frame.html:
```diff
@@ -0,0 +1,10 @@
+<script>
+window.internals.settings.setScreenCaptureEnabled(true);
```
This setting doesn't look like it exists in `InternalSettings`?

bnham replied:
`ScreenCaptureEnabled` is in UnifiedWebPreferences and is also part of the built `JSInternalSettingsGenerated.cpp`?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15624,6 +15636,18 @@ Ref<WebNavigationState> WebPageProxy::protectedNavigationState()
+    RefPtr frame = WebFrameProxy::webFrame(frameID);
+    if (!frame)
+        return nullptr;
+    if (auto* remotePage = m_browsingContextGroup->remotePageInProcess(*this, frame->process()))
```
This needs to be a RefPtr.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15624,6 +15636,18 @@ Ref<WebNavigationState> WebPageProxy::protectedNavigationState()
     return *m_navigationState;
 }
 
+WebProcessProxy* WebPageProxy::processContainingFrameIfExists(std::optional<FrameIdentifier> frameID) const
```
I think WebFrameProxy::process should return the right process for this.

charliewolfe commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -342,25 +356,32 @@ void UserMediaPermissionRequestManagerProxy::finishGrantingRequest(UserMediaPerm
-            m_hasCreatedSandboxExtensionForTCCD = true;
+        if (!m_hasCreatedSandboxExtensionForTCCD.contains(process->coreProcessIdentifier()) && doesPageNeedTCCD(*page)) {
+            handles = SandboxExtension::createHandlesForMachLookup({ "com.apple.tccd"_s }, process->auditToken(), SandboxExtension::MachBootstrapOptions::EnableMachBootstrap);
+            m_hasCreatedSandboxExtensionForTCCD.add(process->coreProcessIdentifier());
```
Maybe `m_hasCreatedSandboxExtensionForTCCD` should be a `WeakHashSet<WebProcessProxy>`, so that its growth is bounded.

eric-carlson commented on LayoutTests/http/tests/site-isolation/resources/getDisplayMedia-starts-frame.html:
```diff
@@ -0,0 +1,10 @@
+});
+streamPromise
+	.then(() => { window.parent.postMessage('success', '*'); })
+	.catch(error => { window.parent.postMessage(`fail (${error})`, '*'); });
```
Ditto

eric-carlson commented on LayoutTests/http/tests/site-isolation/resources/getDisplayMedia-starts-frame.html:
```diff
@@ -0,0 +1,10 @@
+window.internals.settings.setScreenCaptureEnabled(true);
+var streamPromise = null;
+window.internals.withUserGesture(() => {
+	streamPromise = navigator.mediaDevices.getDisplayMedia({ video: true });
```
You also need tests to ensure that `getUserMedia({video:true})` and getUserMedia({audio:true})` work in an isolated iframe

eric-carlson replied:
Nit: tab

bnham committed:
[Site Isolation] getDisplayMedia does not work in a out of process iframes
https://bugs.webkit.org/show_bug.cgi?id=287299
rdar://129194055

Reviewed by Charlie Wolfe and Eric Carlson.

Starting a display stream via getDisplayMedia does not work in an OOP iframe when site isolation is
enabled. This is because many of the related operations assumed that only the main frame's process
could start a capture.

Fix this by making various parts of setting up a display stream (such as `updateCaptureAccess`,
`UserMediaAccessWasGranted`, `UserMediaAccessWasDenied`, extending sandbox to allow capture, etc.)
to use the process associated with the frame in the UserMediaRequest, rather than unconditionally
using the main frame's process.

Since a WebProcessProxy associated with an OOP iframe can enter the process cache while the main
page is still has an active capture session, also make `WebProcessProxy::removeWebPage` check to see
if it's okay to remove capture sandbox extensions when pages are removed from the process.

* LayoutTests/http/tests/site-isolation/mediastream/getDisplayMedia-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getDisplayMedia-starts.html: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-audio-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-audio-starts.html: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-video-starts-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/mediastream/getUserMedia-video-starts.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getDisplayMedia-starts-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getUserMedia-audio-starts-frame.html: Added.
* LayoutTests/http/tests/site-isolation/resources/getUserMedia-video-starts-frame.html: Added.
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::isPlayingMediaDidChange):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::revokeSandboxExtensionsIfNeededForPage):
(WebKit::UserMediaPermissionRequestManagerProxy::disconnectFromPage):
(WebKit::UserMediaPermissionRequestManagerProxy::captureDevicesChanged):
(WebKit::UserMediaPermissionRequestManagerProxy::denyRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::finishGrantingRequest):
(WebKit::UserMediaPermissionRequestManagerProxy::captureStateChanged):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/UserMediaProcessManager.cpp:
(WebKit::UserMediaProcessManager::willCreateMediaStream):
(WebKit::UserMediaProcessManager::revokeSandboxExtensionsIfNeeded):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::userMediaPermissionRequestManagerIfExists):
(WebKit::WebPageProxy::willStartCapture):
(WebKit::WebPageProxy::setOrientationForMediaCapture):
(WebKit::WebPageProxy::gpuProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::removeWebPage):

Canonical link: https://commits.webkit.org/290715@main
